shapeshift/hdwallet
# feat: add bytes field to ethereum signMessage methods

This adds the bytes possibility for messages in `ETHSignMessage` input of `ethSignMessage` methods.
Basically there is no change related to current behaviour with `message` as a string.
But now we are able to pass message as `Bytes` (= `ArrayLike<number>`).

This is required to sign messages for cowswap at least because of the following :
"A common case is to sign a hash. In this case, if the hash is a string, it must be converted to an array first, using the [arrayify](https://docs.ethers.io/v5/api/utils/bytes/#utils-arrayify) utility function."
(from https://docs.ethers.io/v5/api/signer/#Signer-signMessage)

I linked to a custom web in order to test some signatures.
I tested signatures generated from `bytes` input for `Native` and `Metamask` and I was able to get signatures valid for cowswap.
I was not able to get valid signature for Metamask + ledger so probably more work will be needed for this (eip712 implementation ?)

Needed for https://github.com/shapeshift/lib/issues/726

**Testing steps:**

- Build following lib branch with linked hdwallet branch https://github.com/shapeshift/lib/tree/cowswap-executeTrade

- Build following web branch with linked lib / hdwallet https://github.com/shapeshift/web/tree/cowswapper-web-integration-wip . This is a web branch where CowSwapper is used as primary swapper (so it will be used in priority for any erc20 swaps), if you try to trade an ETH pair 0x will be used.
I used local linking syntax `"@shapeshiftoss/hdwallet-core": "../hdwallet/packages/hdwallet-core"` but feel free to do something else.

- Launch web branch locally with native wallet loaded with some ETH / erc20 token (ex : WETH)

- In trade component : trade for example WETH / FOX (erc20 are required to use cowswap) and click on preview trade

- Then click on confirm trade (I have put a glitch in lib so the orders wont be accepted by cowswap , you can have the glitch removed by changing line 46 in `lib/CowExecuteTrade.ts` `validTo: 1656667297, //getNowPlusThirtyMinutesTimestamp()`, use the commented line instead of 1656667297 . Be careful that polling for order completion is not implemented yet on that lib branch).

- If signature is considered valid by cowswap, you should notice a 400 error in network tab for `/orders` api call with following error : `{"errorType":"InsufficientValidTo","description":"validTo is not far enough in the future"}`
If signature is deemed invalid by cowswap, you should receive the following error : `{"errorType":"WrongOwner","description":"Address recovered from signature 0x4xxx…yyy does not match from address"}`

https://api.cow.fi/docs/#/default/post_api_v1_orders can be used to see the various results of the api and verify that by changing the signature we effectively get the mentionned errors.
